### PR TITLE
Admin/laboratory controllers

### PR DESCRIPTION
**Pull request by TheRudno**
Take a look at controlers

### CODE REVIEW

**Comment by pstowski** [comment]
Please avoid using "/" symbol in branch naming.

File: src/main/java/pl/polsl/bd/clinicassistance/services/UserService.java [hunk 1]
```diff
@@ -122,8 +123,28 @@ private String createEncryptedPassword(String password){
...
+    }
 
+    public void saveUserChangesWithoutNewPassword(User user) throws ValidationException, DuplicationException  {
+        validateUserCorrectenss(user);
+        checkDuplicationOfRecords(user);
+        userRepository.save(user);
+    }
 
-
+    public List<Employee> getEmployeesByUserRole(User.EmployeeRole role){
```
Comment (pstowski) [inline src/main/java/pl/polsl/bd/clinicassistance/services/UserService.java]: You should create apropriate method in repository or service - not manually extracting users.
Potential bug: what if some employee has more than one role assigned?

File: src/main/java/pl/polsl/bd/clinicassistance/controllers/AdminAssignEmployeeController.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+public class AdminAssignEmployeeController implements Initializable {
+
+    Employee employee;
+
+    boolean finded = false;
+
+    List<User> users;
+
+    User existingUser;
+
```
Comment (pstowski) [inline src/main/java/pl/polsl/bd/clinicassistance/controllers/AdminAssignEmployeeController.java]: these 4 above should be private

File: src/main/java/pl/polsl/bd/clinicassistance/controllers/AdminAssignEmployeeController.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+import java.util.List;
+import java.util.ResourceBundle;
+
+@Component
+@Scope(scopeName = "prototype")
+public class AdminAssignEmployeeController implements Initializable {
+
+    Employee employee;
+
+    boolean finded = false;
```
Comment (pstowski) [inline src/main/java/pl/polsl/bd/clinicassistance/controllers/AdminAssignEmployeeController.java]: FOUND!!!!

File: src/main/java/pl/polsl/bd/clinicassistance/controllers/ExaminationController.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+import java.net.URL;
+import java.util.ResourceBundle;
+
+@Component
+@Scope(scopeName = "prototype")
+public class ExaminationController implements Initializable {
+
+    LaboratoryExamination examination;
+    User user;
+
```
Comment (pstowski) [inline src/main/java/pl/polsl/bd/clinicassistance/controllers/ExaminationController.java]: Two above should be private

File: src/main/java/pl/polsl/bd/clinicassistance/controllers/ExaminationController.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        textTestCode.setText(examination.getType().getCode());
+        inputResult.setText(examination.getResult());
+        inputRemarks.setText(examination.getSupervisorNote());
+        inputRemarks.setDisable(true);
+        inputResult.setDisable(true);
+        cancelBtn.setDisable(true);
+        confirmBtn.setDisable(true);
+
+        switch (user.getRole()){
+            case LAB_SUPERVIZOR:
```
Comment (pstowski) [inline src/main/java/pl/polsl/bd/clinicassistance/controllers/ExaminationController.java]: Quite nice solution

File: src/main/java/pl/polsl/bd/clinicassistance/controllers/LaboratoryController.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+    @FXML
+    private TableColumn<LaboratoryExamination, String> tabDateOfConf;
+
+    @FXML
+    private TableColumn<LaboratoryExamination, LaboratoryExamination.LaboratoryExaminationStatus> tabStatus;
+
+    @Override
+    public void initialize(URL url, ResourceBundle resourceBundle) {
+        loggedUser = (User) formManager.findSceneContext(this);
```
Comment (pstowski) [inline src/main/java/pl/polsl/bd/clinicassistance/controllers/LaboratoryController.java]: Default filtering should have status = DELEGATED for worker and REALIZED for supervisor

File: src/main/java/pl/polsl/bd/clinicassistance/controllers/LaboratoryController.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        tabDateOfExam.setCellValueFactory(new PropertyValueFactory<>("examinationDate"));
+        tabDateOfConf.setCellValueFactory(new PropertyValueFactory<>("approvalDate"));
+        tabStatus.setCellValueFactory(new PropertyValueFactory<>("status"));
+        examinationList.addAll(examinations);
+        tab.setItems(examinationList);
+        addRowDoubleClickListener();
+    }
+
+    private void addRowDoubleClickListener(){
+        tab.setRowFactory( tv -> {
```
Comment (pstowski) [inline src/main/java/pl/polsl/bd/clinicassistance/controllers/LaboratoryController.java]: it should refresh content after finishing job on Examination screen

File: pom.xml [hunk 8]
```diff
@@ -118,6 +118,13 @@
             <scope>test</scope>
         </dependency>
 
+        <!-- https://mvnrepository.com/artifact/joda-time/joda-time -->
```
Comment (pstowski) [inline pom.xml]: What is this used for?

File: src/main/java/pl/polsl/bd/clinicassistance/controllers/AdminController.java [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+            formManager.createScene(FormManager.Layout.ASSIGN_ROLE,employee,this);
+        }
+    }
+
+    @FXML
+    void onSearchClick(ActionEvent event) {
+        if(roleSearch.getSelectionModel().getSelectedItem()!=null){
+            List<Employee> employeesTmp1 =  userService.getEmployeesByUserRole(User.EmployeeRole.valueOf(roleSearch.getValue().toString()));
+            employees.clear();
+            for (Employee employee: employeesTmp1){
```
Comment (pstowski) [inline src/main/java/pl/polsl/bd/clinicassistance/controllers/AdminController.java]: You should move this logic to service or better to repository

File: pom.xml (hunk 8 above)
Comment (TheRudno) [reply pom.xml]: It is used in utilities to translate this awful timestamp to nice format

File: src/main/java/pl/polsl/bd/clinicassistance/controllers/AdminAssignEmployeeController.java (hunk 2 above)
Comment (TheRudno) [reply src/main/java/pl/polsl/bd/clinicassistance/controllers/AdminAssignEmployeeController.java]: changed

File: src/main/java/pl/polsl/bd/clinicassistance/controllers/AdminAssignEmployeeController.java (hunk 3 above)
Comment (TheRudno) [reply src/main/java/pl/polsl/bd/clinicassistance/controllers/AdminAssignEmployeeController.java]: changed

File: src/main/java/pl/polsl/bd/clinicassistance/controllers/ExaminationController.java (hunk 4 above)
Comment (TheRudno) [reply src/main/java/pl/polsl/bd/clinicassistance/controllers/ExaminationController.java]: changed

**Comment by TheRudno** [comment]
It is "Working controlers, without data validation" so data validation is yet to be added

File: src/main/java/pl/polsl/bd/clinicassistance/services/UserService.java (hunk 1 above)
Comment (TheRudno) [reply src/main/java/pl/polsl/bd/clinicassistance/services/UserService.java]: Potential bug: what if some employee has more than one role assigned? >> It doesn't matter for the controller, employee need to have specific role that's only requirement, if he/she have more, then they will be still displayed in right part of Admin layout panel

Comment (TheRudno) [reply src/main/java/pl/polsl/bd/clinicassistance/services/UserService.java]: Manually extracting users >> I have no idea how to implement this, if you could suggest something I will be glad to try doing it.

File: src/main/java/pl/polsl/bd/clinicassistance/controllers/LaboratoryController.java (hunk 7 above)
Comment (TheRudno) [reply src/main/java/pl/polsl/bd/clinicassistance/controllers/LaboratoryController.java]: I was thinking about it, but how can I determine when the Examination screen is closed to trigger refresh event?

Comment (pstowski) [reply src/main/java/pl/polsl/bd/clinicassistance/controllers/LaboratoryController.java]: There is no way to determine when the examination screen is closed. Nevertheless You can add listener on focus property or something like it, and then if your window or elements regain focus simply refresh results.

Comment (TheRudno) [reply src/main/java/pl/polsl/bd/clinicassistance/controllers/LaboratoryController.java]: done

File: src/main/java/pl/polsl/bd/clinicassistance/controllers/AdminController.java (hunk 9 above)
Comment (TheRudno) [reply src/main/java/pl/polsl/bd/clinicassistance/controllers/AdminController.java]: done

File: src/main/java/pl/polsl/bd/clinicassistance/services/UserService.java (hunk 1 above)
Comment (TheRudno) [reply src/main/java/pl/polsl/bd/clinicassistance/services/UserService.java]: done

File: src/main/java/pl/polsl/bd/clinicassistance/controllers/AdminController.java [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+    @FXML
+    void onModifyRolesClick(ActionEvent event) {
+        if(employeeTable.getSelectionModel().getSelectedItem() != null){
+            Employee employee = employeeTable.getSelectionModel().getSelectedItem();
+            formManager.createScene(FormManager.Layout.ASSIGN_ROLE,employee,this);
+        }
+    }
+
+    @FXML
+    void onSearchClick(ActionEvent event) {
```
[the remark this answers is not in the document]
Comment (TheRudno) [reply src/main/java/pl/polsl/bd/clinicassistance/controllers/AdminController.java]: done

File: src/main/java/pl/polsl/bd/clinicassistance/controllers/LaboratoryController.java (hunk 6 above)
Comment (TheRudno) [reply src/main/java/pl/polsl/bd/clinicassistance/controllers/LaboratoryController.java]: done

**Pull request closed without merging**